### ISSUE TITLE
seq[T] does not respect alignof(T)

### ISSUE THREAD

**Issue by jxy** [nim-lang/Nim#14642]
The data in an `seq[T]` does not have the proper alignment as `alignof(T)`.

### Example
```nim
type
  VFloat = object
    d {.align(64).}: array[8,float]
  VData[T] = ref object
    s: seq[T]

proc main =
  var
   f = VData[float](s: newseq[float](64))
   v = VData[VFloat](s: newseq[VFloat](8))
  echo "sizeof(f.s[0]) = ",sizeof(f.s[0])
  echo "alignof(f.s[0]) = ",alignof(f.s[0])
  echo "addr(f.s[0]) = ",repr(addr(f.s[0]))
  echo "sizeof(v.s[0]) = ",sizeof(v.s[0])
  echo "alignof(v.s[0]) = ",alignof(v.s[0])
  echo "addr(v.s[0]) = ",repr(addr(v.s[0]))

main()
```

### Current Output
```
sizeof(f.s[0]) = 8
alignof(f.s[0]) = 8
addr(f.s[0]) = ptr 0x10dfec060 --> 0.0

sizeof(v.s[0]) = 64
alignof(v.s[0]) = 64
addr(v.s[0]) = ptr 0x10dfed090 --> [d = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]

```
The address of `v.s[0]` is not properly aligned.

**Comment by planetis-m** [comment]
I think the only way to make this work properly, is to move the ``data`` field in ``NimSeqV2``. like this:

```nim
type
  NimSeqPayload = object
    cap: int

  NimSeqV2*[T] = object
    len: int
    data: ptr UncheckedArray[T]
    p: ptr NimSeqPayload

proc newSeqPayload(cap: int): ptr NimSeqPayload =
  var p = cast[ptr NimSeqPayload](allocShared0(align(sizeof(NimPayload), alignof(T)) + cap * sizeof(T)))
  p.cap = cap
  result = p

proc newSeq[T](cap: Natural): NimSeqV2[T] =
   result = NimSeqV2[T](len: 0, p: newSeqPayload[T](cap))
   result.data = cast[typeof(result.data)](result.p +! align(sizeof(NimSeqPayload), alignof(T)))
```
Edit: nope my code is wrong.

**Comment by shirleyquirk** [comment]
this has nothing to do with codegen or pragmas, or even about seq in particular, it's about achieving alignment on a heap allocation.

```nim
import macros
type
  VFloat = object
    i: int8
    d {.align(64).}: array[8, float]
macro info(x: untyped) =
  let xrep = x.repr
  quote do:
    echo "sizeof(", `xrep`, ") = ", sizeof(`x`)
    echo "alignof(", `xrep`, ") = ", alignof(`x`)
    echo "addr(", `xrep`, ") = ", repr(addr(`x`))
    echo "addr  mod alignof = ", cast[int](addr(`x`)) mod alignof(`x`),'\n'
proc main =
  var
    stack = VFloat()
    heap = new VFloat
  heap[] = stack
  info(heap[])
  info(stack)

main()
```
produces:
```
sizeof(heap[]) = 128
alignof(heap[]) = 64
addr(heap[]) = ptr 0x7f2d0d028050 --> [i = 0,
d = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]

addr mod alignof = 16

sizeof(stack) = 128
alignof(stack) = 64
addr(stack) = ptr 0x7ffdbb9fb240 --> [i = 0,
d = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]]

addr mod alignof = 0
```

for the malloc using gc's:
   one solution is to use ``aligned_alloc`` which is  C11 or std::aligned_alloc in C++17, not sure in objc

On embedded it's chaotic, some have it, some dont, some have ``memalign`` or ``posix_memalign``  esp-idf has ``memalign`` and forwards it to an error message.

the do it yourself pointermath solution is malloc alignof(T)-1 more memory than you need, shift up to the proper alignment, store the offset in the previous byte(s), return the shifted pointer.  aligned-free(p) calls free(p-(p-1)[]) 

this would get messy if there were two kinds of allocation going on, dealloc'ing an aligned pointer would either crash or leak memory or both, aligned_dealloc'ing a non-aligned pointer, crashing would probably be the best case scenario.

https://github.com/nim-lang/Nim/pull/12430 did a lot of this work and
https://github.com/nim-lang/Nim/pull/12926 continues it

**Comment by planetis-m** [comment]
Fixed in #15697!!
